### PR TITLE
Adding FGAC annotations validation for creation and deletion

### PR DESCRIPTION
**Pull request by evictorero**
**What this PR does / why we need it**:
* adding new actions for create and delete annotations
* adding FGAC auth checks to creation endpoint
* adding FGAC auth checks to creation graphite annotations endpoint
* adding FGAC auth checks to deletion annotations endpoint
* adding tests to cover all new auth checks plus PUT and PATCH endpoints
* refactor naming `local` and `global` annotations in favor by `dashboard` and `organization` annotations

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/grafana-enterprise/issues/2999

**Special notes for your reviewer**:

Creation endpoint requires additional validation at handler level because it is necessary to look into the body to determine if it is a local or global annotation. Graphite endpoint creation does not require such validation because it is always global.
Deletion endpoint FGAC auth is covered in the middleware

### CODE REVIEW

File: pkg/api/accesscontrol.go [hunk 1]
```diff
@@ -276,15 +276,15 @@ func (hs *HTTPServer) declareFixedRoles() error {
...
-	localAnnotationsWriterRole := ac.RoleRegistration{
+	dashboardAnnotationsWriterRole := ac.RoleRegistration{
 		Role: ac.RoleDTO{
-			Name:        "fixed:annotations.local:writer",
-			DisplayName: "Local annotation writer",
+			Name:        "fixed:annotations.dashboard:writer",
+			DisplayName: "Dashboard annotation writer",
 			Description: "Update annotations associated with dashboards.",
 			Group:       "Annotations",
 			Version:     1,
```
Comment (IevaVasiljeva) [inline pkg/api/accesscontrol.go:285]: Commenting so that we don't forget: need to increase the version here :)

File: pkg/api/annotations.go [hunk 2]
```diff
@@ -399,11 +411,16 @@ func AnnotationTypeScopeResolver() (string, accesscontrol.AttributeScopeResolveF
...
 		} else {
-			return accesscontrol.ScopeAnnotationsTypeLocal, nil
+			return accesscontrol.ScopeAnnotationsTypeDashboard, nil
 		}
 	}
 	return accesscontrol.ScopeAnnotationsProvider.GetResourceScope(""), annotationTypeResolver
 }
+
+func (hs *HTTPServer) canCreateOrganizationAnnotation(c *models.ReqContext) (bool, error) {
+	evaluator := accesscontrol.EvalPermission(accesscontrol.ActionAnnotationsWrite, accesscontrol.ScopeAnnotationsTypeOrganization)
```
Comment (IevaVasiljeva) [inline pkg/api/annotations.go:424]: So far we've been adding separate actions for resource creation, updating and deletion. So we should probable introduce a new `annotations:create` action and use that for this evaluation.

File: pkg/api/annotations_test.go [hunk 3]
```diff
@@ -458,10 +485,217 @@ func TestAPI_Annotations_AccessControl(t *testing.T) {
 			},
 			want: http.StatusForbidden,
 		},
+		{
```
Comment (IevaVasiljeva) [inline pkg/api/annotations_test.go:488]: Really nice tests, thanks for adding them all!

File: pkg/api/api.go [hunk 4]
```diff
@@ -440,11 +440,11 @@ func (hs *HTTPServer) registerRoutes() {
 		apiRoute.Post("/annotations/mass-delete", reqOrgAdmin, routing.Wrap(hs.DeleteAnnotations))
 
 		apiRoute.Group("/annotations", func(annotationsRoute routing.RouteRegister) {
-			annotationsRoute.Post("/", routing.Wrap(hs.PostAnnotation))
-			annotationsRoute.Delete("/:annotationId", routing.Wrap(hs.DeleteAnnotationByID))
+			annotationsRoute.Post("/", authorize(reqSignedIn, ac.EvalPermission(ac.ActionAnnotationsWrite)), routing.Wrap(hs.PostAnnotation))
+			annotationsRoute.Delete("/:annotationId", authorize(reqSignedIn, ac.EvalPermission(ac.ActionAnnotationsWrite, ac.ScopeAnnotationsID)), routing.Wrap(hs.DeleteAnnotationByID))
```
Comment (IevaVasiljeva) [inline pkg/api/api.go:444]: Similarly as for creation, I think we should add a new `annotations:delete` action for this permission evaluation.

**Review by IevaVasiljeva — APPROVED** [review]
Looks great, thanks for all the work!

The only thing I think needs changing is introducing new actions for annotation creation and deletion, using those for the corresponding permission checks, as well as adding them to the fixed role grants in `pkg/api/accesscontrol.go`. Apart from that the scopes, tests and everything looks good 👌

File: pkg/api/accesscontrol.go (hunk 1 above)
Comment (evictorero) [reply pkg/api/accesscontrol.go]: Done thanks!

File: pkg/api/api.go (hunk 4 above)
Comment (evictorero) [reply pkg/api/api.go]: Done!

File: pkg/api/annotations.go (hunk 2 above)
Comment (evictorero) [reply pkg/api/annotations.go]: Done!

**Review by IevaVasiljeva — APPROVED** [review]
Nice! 🚢

**Review by IevaVasiljeva — APPROVED** [review]

**Pull request merged**